force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,63 +211,154 @@ public with sharing class DMLOperations {
+     * If no such Account exists, it creates a new one and sets the description to "New Account".
+     * @param accountName The name of the Account to be upserted.
+     * @return The Account that was upserted.
+     */
+    public static Account upsertAccount(String accountName) {
+        
+        // Query for the Account
+        List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
+
+
+        Account accountToUpsert;
+
+        // if the Account exists, update the description
+
+        if (!accounts.isEmpty()) {
+            accountToUpsert = accounts[0]   ;
+            accountToUpsert.Description = 'Updated Account' ;
+            
+        }
 
-		// if the Account exists, update the description
+        // otherwise, create a new Account and set the description
+
+        else {
+            accountToUpsert = new Account(Name = accountName, Description = 'New Account')  ;
+        }
+
+        // Upsert the Account
+
+        upsert accountToUpsert  ;
+
+        return accountToUpsert  ;
+    }
 
-		// otherwise, create a new Account and set the description
 
-		// Upsert the Account
-		return null;
-	}
 
     /**
      * Question 10
-	 * Manage the association between Contacts and Accounts
-	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+     * Manage the association between Contacts and Accounts
+     * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
+     * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - 
+     * this may cause bad practices but for learning purposes it's okay.)
      * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
-	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
-	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
-		// Loop through each Contact
+     * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
+     */
+    public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+
+        // Loop through each Contact
+        for (Contact cont : contactsToUpsert) {
+
+            // Extract the Account Name from Contact's LastName
+            String accountName = cont.LastName;
+
+            // Get the account based on the name or create one if it doesn't exist
+            Account accountToUpsert = upsertAccount(accountName)    ;
```

Comment:
Good job using the previous `accountToUpsert` method. It saves you a lot of work!